File: 0054-spiral-matrix/0054-spiral-matrix.cpp
// TC- O(m*n)
// SC- O(1)

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> ans;

        int left=0;
        int up=0;
        int down = matrix.size()-1;
        int right=matrix[0].size()-1;

        while(ans.size()<(matrix.size()*matrix[0].size())){
            for(int col=left; col<=right; col++){
                ans.push_back(matrix[up][col]);
            }
            for(int row=up+1; row<=down; row++){
                ans.push_back(matrix[row][right]);
            }

            // Check if it is a different row before you start traversing right->left
            if(up!=down){
                for(int col=right-1; col>=left; col--){
                    ans.push_back(matrix[down][col]);
                }
            }
            // Check if it is a different col before you start traversing down->up
            if(left!=right){
                for(int row=down-1; row>up; row--){
                    ans.push_back(matrix[row][left]);
                }
            }

            // Update limits
            left++; right--; up++; down--;
        }
        return ans;
    }
};

File: 0054-spiral-matrix/0054-spiral-matrix.py
# TC- O(mn)
# SC- O(1)

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows, cols = len(matrix), len(matrix[0])
        ans = []

        left, right, top, bottom = 0, cols-1, 0, rows-1
        while(len(ans)!=(rows*cols)):
            for c in range(left, right+1):
                ans.append(matrix[top][c])
            top += 1

            for r in range(top, bottom+1):
                ans.append(matrix[r][right])
            right -= 1

            if top<= bottom:
                for c in range(right, left-1, -1):
                    ans.append(matrix[bottom][c])
                bottom -= 1
            
            if left <= right:
                for r in range(bottom, top-1, -1):
                    ans.append(matrix[r][left])
                left += 1

        return ans